File: 0079_Word Search/WordSearch.cs
public class Solution {
    public bool Exist(char[][] board, string word) {
        if(board == null) return false;
        for(int i=0;i<board.Length;i++){
            for(int j=0;j<board[i].Length;j++){
                if(Search(board,word,i,j,0)) return true;        
            }
        }
        
        return false;
    }
    
    private bool Search(char[][] board,string word, int x,int y, int p){
        if(x<0 || y<0 || x==board.Length || y==board[x].Length || board[x][y] != word[p]) return false;
        //not necessary using this,since the next line always returns false if board[x][y] is '#'
        //if(board[x][y] == '#') return false;
        if(p == word.Length-1) return true;
        char original = board[x][y];
        board[x][y] = '#';
        var result =   Search(board, word, x-1,y,p+1) ||
                       Search(board, word, x+1,y,p+1) ||
                       Search(board, word, x,y-1,p+1) ||
                       Search(board, word, x,y+1,p+1);
        board[x][y] = original;
        return result;
        }
}

File: 0079_Word Search/WordSearch.java
class Solution {
    public boolean exist(char[][] board, String word) {
        int m = board.length, n = board[0].length;
        int[] dirs = {0,1,0,-1,0};
        for(int i=0;i<m;i++) {
            for(int j=0;j<n;j++)
                if(DFS(board, i, j, m, n, 0, word, dirs)) return true;
        }
        
        return false;
    }
    
    private boolean DFS(char[][] board, int x, int y, int m, int n, int index, String word, int[] dirs) {
        if(index == word.length()) return true;
        if(x < 0 || y < 0 || x == m || y == n || board[x][y] != word.charAt(index)) return false;
        char c = board[x][y];
        board[x][y] = '#';
        for(int i=0;i<4;i++) {
            int dx = x + dirs[i], dy = y + dirs[i+1];
            if(DFS(board, dx, dy, m, n, index+1, word, dirs)) return true;
        }
        board[x][y] = c;
        
        return false;
    }
}